tests/interfaces/test_batch_jax.py

```diff
@@ -0,0 +1,578 @@
+        # for the forward pass, and one for the backward pass.
+        dev._num_executions = 0
+        jac_fn = jax.grad(cost)
+        grad1 = jac_fn(params, cache=True)
+        assert dev.num_executions == 2
+
+
+execute_kwargs = [
+    {"gradient_fn": param_shift},
+    {
+        "gradient_fn": "device",
+        "mode": "forward",
+        "gradient_kwargs": {"method": "adjoint_jacobian", "use_device_state": True},
+    },
+    {
+        "gradient_fn": "device",
+        "mode": "backward",
+        "gradient_kwargs": {"method": "adjoint_jacobian"},
+    },
+]
+
+
+@pytest.mark.parametrize("execute_kwargs", execute_kwargs)
+class TestJaxExecuteIntegration:
+    """Test the jax interface execute function
+    integrates well for both forward and backward execution"""
+
+    def test_execution(self, execute_kwargs):
+        """Test execution"""
+        dev = qml.device("default.qubit", wires=1)
+
+        def cost(a, b):
+            with qml.tape.JacobianTape() as tape1:
+                qml.RY(a, wires=0)
+                qml.RX(b, wires=0)
+                qml.expval(qml.PauliZ(0))
+
+            with qml.tape.JacobianTape() as tape2:
+                qml.RY(a, wires=0)
+                qml.RX(b, wires=0)
+                qml.expval(qml.PauliZ(0))
+
+            return execute([tape1, tape2], dev, interface="jax", **execute_kwargs)
+
+        a = jnp.array(0.1)
+        b = jnp.array(0.2)
+        res = cost(a, b)
+
+        assert len(res) == 2
+        assert res[0].shape == (1,)
+        assert res[1].shape == (1,)
+
+    def test_scalar_jacobian(self, execute_kwargs, tol):
+        """Test scalar jacobian calculation"""
+        a = jnp.array(0.1)
+        dev = qml.device("default.qubit", wires=2)
+
+        def cost(a):
+            with qml.tape.JacobianTape() as tape:
+                qml.RY(a, wires=0)
+                qml.expval(qml.PauliZ(0))
+            return execute([tape], dev, interface="jax", **execute_kwargs)[0][0]
+
+        res = jax.grad(cost)(a)
+        assert res.shape == ()
+
+        # compare to standard tape jacobian
+        with qml.tape.JacobianTape() as tape:
+            qml.RY(a, wires=0)
+            qml.expval(qml.PauliZ(0))
+
+        tape.trainable_params = {0}
+        tapes, fn = param_shift(tape)
+        expected = fn(dev.batch_execute(tapes))
+
+        assert expected.shape == (1, 1)
+        assert np.allclose(res, np.squeeze(expected), atol=tol, rtol=0)
+
+    def test_reusing_quantum_tape(self, execute_kwargs, tol):
+        """Test re-using a quantum tape by passing new parameters"""
+        a = jnp.array(0.1)
+        b = jnp.array(0.2)
+
+        dev = qml.device("default.qubit", wires=2)
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RY(a, wires=0)
+            qml.RX(b, wires=1)
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0))
+
+        assert tape.trainable_params == {0, 1}
+
+        def cost(a, b):
+            tape._update()

```

Comment:
@josh146 this was an update I had to add here for JAX.

It seems that with calling `_update`, the trainable parameters are incorrect and an error is raised here:
https://github.com/PennyLaneAI/pennylane/blob/6d93cc580d4e8860a0b3fc8cc42e81198399e665/pennylane/tape/tape.py#L861

Reply:
got it :+1:

Might be worth adding this as a comment here, to note that this differs from autograd, and to also point out the failing line above?

Reply:
Sounds good, done!